### PR TITLE
CDAT migration: Fix African easterly wave density plots in TC analysis and convert H20LNZ units to ppm/volume

### PR DESCRIPTION
**Pull request by tomvothecoder**
## Description

- Related to #851 -- looks like I already addressed this on `cdat-migration-fy24` but I refactored it a bit more
- Related to #874 
- Relatedt to #871 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

**Review by tomvothecoder — COMMENTED** [review]
Hey @chengzhuzhang this PR is ready for review. I've performed a self-review, but good to make sure the changes align with your two PRs mentioned in this PR's description.

Summary of changes
- `tc_analysis_driver.py` 
   - Remove `lat`/`lon` slice
   - Add function `_filter_lines_within_year_bounds()` to remove excessive itme points crossing year bounds for 6 hourly data
 - `zonal_mean_2d_driver.py`
     - Update `_convert_g_kg_to_ppm_units()` to convert to ppm by volume.

File: e3sm_diags/driver/tc_analysis_driver.py [hunk 1]
```diff
@@ -243,6 +230,43 @@ def generate_tc_metrics_from_te_stitch_file(te_stitch_file: str) -> Dict[str, An
...
+    lines_orig : List[str]
+        A list of strings where each string represents a line of data.
+    data_end_year : int
+        The end year for filtering lines. Only lines with years less than or
+        equal to this value will be retained.
+    Returns
+    -------
+    List[str]
+        A list of strings filtered based on the specified year bounds.
+    """
```
Comment (tomvothecoder) [inline e3sm_diags/driver/tc_analysis_driver.py:255]: FYI I used GitHub Co-Pilot to extract this function and generation the docstring.

File: e3sm_diags/driver/tc_analysis_driver.py [hunk 2]
```diff
@@ -243,6 +230,43 @@ def generate_tc_metrics_from_te_stitch_file(te_stitch_file: str) -> Dict[str, An
...
+        if lines_orig[i][0] == "s":
+            year = int(lines_orig[i].split("\t")[2])
+
+            if year <= data_end_year:
+                line_ind.append(i)
+
+    end_ind = line_ind[-1]
+
+    new_lines = lines_orig[0:end_ind]
+    return new_lines
```
Comment (tomvothecoder) [inline e3sm_diags/driver/tc_analysis_driver.py:267]: FYI I used GitHub Co-Pilot to extract this function and generated the docstring. I've found it to be useful for these kinds of tasks.

Comment (chengzhuzhang) [reply e3sm_diags/driver/tc_analysis_driver.py:267]: Nice finding!

File: e3sm_diags/driver/zonal_mean_2d_driver.py [hunk 3]
```diff
@@ -161,7 +161,7 @@ def _convert_g_kg_to_ppm_units(
             parameter.current_set == "zonal_mean_2d_stratosphere"
             and parameter.var_id == "Q"
         ):
-            ds_new[var_key] = ds_new[var_key] * 1000.0
+            ds_new[var_key] = ds_new[var_key] * 28.97 / 18.0 * 1000.0
             ds_new[var_key].attrs["units"] = "ppm"
```
Comment (chengzhuzhang) [inline e3sm_diags/driver/zonal_mean_2d_driver.py:165]: The new units should be "ppmv" with "v" indicating ppm by volume.

**Comment by tomvothecoder** [comment]
> @tomvothecoder I think the changes in .cfg files for H2OLNZ units change are not brought in this PR: [#874 (files)](https://github.com/E3SM-Project/e3sm_diags/pull/874/files#diff-794c0344b0a718e616d2051d6961d02c79f24e8d5e97c7c355fa7cb0ec467e19)

For that specific PR, the changes are already here. Example: https://github.com/E3SM-Project/e3sm_diags/blob/3df16d2266c98f526fe03ebeecb68c6170d61618/e3sm_diags/driver/default_diags/zonal_mean_2d_stratosphere_model_vs_obs.cfg#L65-L75

I rebased the dev branch on the latest `main`, which pulls in changes to `.cfg` files. I don't need to update any `.cfg` files in this PR.

File: e3sm_diags/driver/zonal_mean_2d_driver.py (hunk 3 above)
Comment (tomvothecoder) [reply e3sm_diags/driver/zonal_mean_2d_driver.py:165]: Thanks for the catch, just updated.

Comment (tomvothecoder) [inline e3sm_diags/driver/zonal_mean_2d_driver.py:165]:
```suggestion
            ds_new[var_key].attrs["units"] = "ppmv"
```

**Review by chengzhuzhang — APPROVED** [review]
Looking good! The PR is good to go!

**Comment by chengzhuzhang** [comment]
@tomvothecoder Just a heads-up that this PR is also just merged: https://github.com/E3SM-Project/e3sm_diags/pull/879  and we need to bring to migration branch..

**Comment by tomvothecoder** [comment]
> @tomvothecoder Just a heads-up that this PR is also just merged: #879 and we need to bring to migration branch..

Done

**Pull request merged**